### PR TITLE
enip: convert to rust

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3958

Describe changes:
- convert enip parser to rust

Alon the way, also
- transactions are now bidirectional
- there is a enip logger
- gap support is improved with probing for resync
- frames
- events
- enip_command keyword accepts now string enumeration as values.
- more keywords, to have parity with logging

#10850 with needed rebase

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1666

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_autofp_suri_time.

ERROR: QA failed on SURI_TLPW1_files_sha256.

ERROR: QA failed on SURI_TLPR1_suri_time.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 138 | 145 | 105.07% |
| | SURI_TLPR1_stats_chk | | |
| .uptime | 644 | 693 | 107.61% | 

Pipeline 20220

**Review by jufajardini — COMMENTED** [review]
Kudos on tackling keyword/log output parity :D

Not sure if possible, but wondering if we could have dedicated commits:
Maybe:
- one for docs
- one for logging and parsing (and schema?)
- one for detection
- one for new keywords added
- replacement of C code with rust code into an independent commit
- maybe one for unittests removal, just to keep the change contained?

Please ignore if this proposal doesn't make sense or would be too much work

**Comment by catenacyber** [comment]
> Not sure if possible, but wondering if we could have dedicated commits: Maybe:
> 
> * one for docs

I do not like separating doc commits from the code changes needing the documentation...
I do not think this was settled

> * one for logging and parsing (and schema?)
> * one for detection

Already existing keywords must be done within the same commit as parsing goes from c to rust...

> * one for new keywords added
> * replacement of C code with rust code into an independent commit

This is not independent : I add logging and parsing in rust, while moving away the C code...

> * maybe one for unittests removal, just to keep the change contained?

Interesting. Why so ?

> 
> Please ignore if this proposal doesn't make sense or would be too much work

This can be work... not sure if it is relevant...

**Comment by jufajardini** [comment]
> > * maybe one for unittests removal, just to keep the change contained?
> 
> Interesting. Why so ?

Looking again, I can't see what made me make this comment, please ignore.  >__<'

> 
> > Please ignore if this proposal doesn't make sense or would be too much work
> 
> This can be work... not sure if it is relevant...

In general I see that we try to break changes into smaller commits, so that's where this comes from. But I don't have a super strong opinion about this, yet...

File: rust/src/enip/logger.rs [hunk 1]
```diff
@@ -0,0 +1,1901 @@
...
+    } else {
+        js.set_string("status", &format!("unknown-{}", h.status))?;
+    }
+    if h.options != 0 {
+        js.set_uint("options", h.options.into())?;
+    }
+    Ok(())
+}
+
+fn enip_vendorid_string(p: u16) -> Option<&'static str> {
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/enip/logger.rs:46]: https://github.com/wireshark/wireshark/blob/master/epan/dissectors/packet-cip.c#L1579

[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/enip/logger.rs:46]: Or https://github.com/cisagov/icsnpp-enip/blob/main/scripts/consts.zeek#L325

File: rust/src/enip/logger.rs [hunk 2]
```diff
@@ -0,0 +1,1901 @@
...
+        1509 => Some("Rethink Robotics, Inc."),
+        1510 => Some("Ingeteam"),
+        1511 => Some("TOSEI ENGINEERING CORP."),
+        1512 => Some("SAMSON AG"),
+        1513 => Some("TGW Mechanics GmbH"),
+        _ => None,
+    }
+}
+
+fn enip_devicetype_string(p: u16) -> Option<&'static str> {
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/enip/logger.rs:1359]: There is also https://github.com/secdev/scapy/blob/master/scapy/contrib/enipTCP.py#L64

File: rust/src/enip/enip.rs [hunk 3]
```diff
@@ -0,0 +1,658 @@
...
+        None
+    }
+
+    fn parse_udp(
+        &mut self, stream_slice: StreamSlice, request: bool, flow: *const Flow,
+    ) -> AppLayerResult {
+        let input = stream_slice.as_slice();
+        match parser::parse_enip_pdu(input) {
+            Ok((_, pdu)) => {
+                process_frames(&pdu, &stream_slice, flow, input);
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/enip/enip.rs:219]: Do you mean only for UDP ?

And which frame are you talking about ? There are many...
I think you want a frame `pdu` which is both the `header` and `payload` frame, right ?

**Comment by catenacyber** [comment]
Continued in https://github.com/OISF/suricata/pull/11174

File: rust/src/enip/detect.rs [hunk 4]
```diff
@@ -0,0 +1,726 @@
...
+                    return enip_cip_match_attribute(&d.cip, ctx);
+                }
+            }
+        }
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_enip_tx_has_cip_attribute(
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/enip/detect.rs:449]: With python
```
f = open("diff.txt")
for l in f.readlines():
    base = l[:-1]
    trans1 = base.replace("rs_enip", "SCEnip")
    trans = ""
    state = 0
    for c in trans1:
        if c == '_':
            state = 1
        else:
            if state == 1:
                trans += c.upper()
                state = 0
            else:
                trans += c
    print("git grep %s | cut -d: -f1 | xargs sed -i -e 's/%s/%s/'" % (base, base, trans))
```

**Pull request closed without merging**